File: app/src/main/java/me/ycdev/android/devtools/apps/installed/InstalledAppsActivity.java
package me.ycdev.android.devtools.apps.installed;

import android.app.Activity;
import android.os.AsyncTask;
import android.os.Bundle;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.widget.AdapterView;
import android.widget.ListView;

import java.util.List;

import me.ycdev.android.arch.activity.AppCompatBaseActivity;
import me.ycdev.android.devtools.R;
import me.ycdev.android.lib.common.apps.AppInfo;
import me.ycdev.android.lib.common.apps.AppsLoadConfig;
import me.ycdev.android.lib.common.apps.AppsLoadFilter;
import me.ycdev.android.lib.common.apps.AppsLoadListener;
import me.ycdev.android.lib.common.apps.AppsLoader;
import me.ycdev.android.lib.commonui.base.LoadingAsyncTaskBase;
import timber.log.Timber;

public class InstalledAppsActivity extends AppCompatBaseActivity
        implements AdapterView.OnItemClickListener {
    private static final String TAG = "InstalledAppsActivity";

    private InstalledAppsAdapter mAdapter;
    private AppsLoadingTask mAppsLoader;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        setContentView(R.layout.installed_apps);
        ListView listView = (ListView) findViewById(R.id.list);
        listView.setOnItemClickListener(this);
        mAdapter = new InstalledAppsAdapter(this);
        listView.setAdapter(mAdapter);

        loadApps();
    }

    private void loadApps() {
        if (mAppsLoader != null && mAppsLoader.getStatus() != AsyncTask.Status.FINISHED) {
            mAppsLoader.cancel(true);
        }
        mAppsLoader = new AppsLoadingTask(this);
        mAppsLoader.execute();
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        MenuInflater inflater = getMenuInflater();
        inflater.inflate(R.menu.installed_apps_menu, menu);
        return super.onCreateOptionsMenu(menu);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        int id = item.getItemId();
        if (id == R.id.action_sort_by_app_name) {
            mAdapter.sort(new AppInfo.AppNameComparator());
            item.setChecked(true);
            return true;
        } else if (id == R.id.action_sort_by_pkg_name) {
            mAdapter.sort(new AppInfo.PkgNameComparator());
            item.setChecked(true);
            return true;
        } else if (id == R.id.action_sort_by_uid) {
            mAdapter.sort(new AppInfo.UidComparator());
            item.setChecked(true);
            return true;
        } else if (id == R.id.action_sort_by_install_time) {
            mAdapter.sort(new AppInfo.InstallTimeComparator());
            item.setChecked(true);
            return true;
        } else if (id == R.id.action_sort_by_update_time) {
            mAdapter.sort(new AppInfo.UpdateTimeComparator());
            item.setChecked(true);
            return true;
        }
        return super.onOptionsItemSelected(item);
    }

    @Override
    public void onItemClick(AdapterView<?> parent, View view, int position, long id) {
        AppInfo item = mAdapter.getItem(position);
        Timber.tag(TAG).i("clicked item: %s", item.toString());
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        if (mAppsLoader != null && mAppsLoader.getStatus() != AsyncTask.Status.FINISHED) {
            mAppsLoader.cancel(true);
        }
    }

    private class AppsLoadingTask extends LoadingAsyncTaskBase<Void, List<AppInfo>> {
        public AppsLoadingTask(Activity cxt) {
            super(cxt);
        }

        @Override
        protected List<AppInfo> doInBackground(Void... params) {
            AppsLoadFilter filter = new AppsLoadFilter();
            filter.setOnlyMounted(false);
            filter.setOnlyEnabled(false);

            AppsLoadConfig config = new AppsLoadConfig();

            AppsLoadListener listener = new AppsLoadListener() {
                @Override
                public boolean isCancelled() {
                    return AppsLoadingTask.this.isCancelled();
                }

                @Override
                public void onProgressUpdated(int percent, AppInfo appInfo) {
                    publishProgress(percent);
                }
            };

            return AppsLoader.Companion.getInstance(getActivity()).loadInstalledApps(filter, config, listener);
        }

        @Override
        protected void onPostExecute(List<AppInfo> result) {
            super.onPostExecute(result);
            mAdapter.setData(result);
            mAdapter.sort(new AppInfo.AppNameComparator());
        }
    }
}

File: app/src/main/java/me/ycdev/android/devtools/apps/installed/InstalledAppsActivity.kt
package me.ycdev.android.devtools.apps.installed

import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import android.view.View
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProvider
import androidx.recyclerview.widget.DividerItemDecoration
import androidx.recyclerview.widget.LinearLayoutManager
import me.ycdev.android.arch.activity.AppCompatBaseActivity
import me.ycdev.android.devtools.R
import me.ycdev.android.devtools.databinding.ActCommonListBinding
import me.ycdev.android.lib.common.apps.AppInfo.AppNameComparator
import me.ycdev.android.lib.common.apps.AppInfo.InstallTimeComparator
import me.ycdev.android.lib.common.apps.AppInfo.MinSdkComparator
import me.ycdev.android.lib.common.apps.AppInfo.PkgNameComparator
import me.ycdev.android.lib.common.apps.AppInfo.TargetSdkComparator
import me.ycdev.android.lib.common.apps.AppInfo.UidComparator
import me.ycdev.android.lib.common.apps.AppInfo.UpdateTimeComparator
import timber.log.Timber

class InstalledAppsActivity : AppCompatBaseActivity() {
    private lateinit var binding: ActCommonListBinding

    private lateinit var listAdapter: InstalledAppsAdapter

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActCommonListBinding.inflate(layoutInflater)
        setContentView(binding.root)

        setSupportActionBar(binding.toolbar)
        supportActionBar?.setDisplayHomeAsUpEnabled(true)

        listAdapter = InstalledAppsAdapter(this)
        binding.list.apply {
            adapter = listAdapter
            layoutManager = LinearLayoutManager(this@InstalledAppsActivity)
            addItemDecoration(
                DividerItemDecoration(this@InstalledAppsActivity, DividerItemDecoration.VERTICAL)
            )
        }

        val viewModel = ViewModelProvider(this).get(InstalledAppsViewModel::class.java)
        viewModel.apps.observe(this, Observer {
            Timber.tag(TAG).d("apps loaded: ${it.size}")
            listAdapter.data = it
            listAdapter.sort(AppNameComparator())
            binding.progress.visibility = View.GONE
        })
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        val inflater = menuInflater
        inflater.inflate(R.menu.installed_apps_menu, menu)
        return super.onCreateOptionsMenu(menu)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.action_sort_by_app_name -> {
                listAdapter.sort(AppNameComparator())
                item.isChecked = true
                return true
            }
            R.id.action_sort_by_pkg_name -> {
                listAdapter.sort(PkgNameComparator())
                item.isChecked = true
                return true
            }
            R.id.action_sort_by_uid -> {
                listAdapter.sort(UidComparator())
                item.isChecked = true
                return true
            }
            R.id.action_sort_by_install_time -> {
                listAdapter.sort(InstallTimeComparator())
                item.isChecked = true
                return true
            }
            R.id.action_sort_by_update_time -> {
                listAdapter.sort(UpdateTimeComparator())
                item.isChecked = true
                return true
            }
            R.id.action_sort_by_target_sdk -> {
                listAdapter.sort(TargetSdkComparator())
                item.isChecked = true
                return true
            }
            R.id.action_sort_byt_min_sdk -> {
                listAdapter.sort(MinSdkComparator())
                item.isChecked = true
                return true
            }
            else -> return super.onOptionsItemSelected(item)
        }
    }

    companion object {
        private const val TAG = "InstalledAppsActivity"
    }
}
